source/gems/dry-validation/comparison-with-activemodel.html.md

```diff
@@ -0,0 +1,606 @@
+validates :attr, numericality: { odd: true }
+```
+
+**dry-validation**
+
+```ruby
+Dry::Validation.Schema do
+  required(:attr) { int? & odd? }
+end
+```
+
+#### Options - even
+
+**ActiveModel Validation**
+
+```ruby
+validates :attr, numericality: { even: true }
+```
+
+**dry-validation**
+
+```ruby
+Dry::Validation.Schema do
+  required(:attr).filled(:int?, :even?)
+end
+```
+
+> Note: `odd?` and `even?` predicates can only be used on integers.
+
+**Additional Uses:**
+
+dry-validation's predicates uses basic Ruby equality operators (`<`, `>`, `==` etc.) which means that they can be used to validate anything that's comparable.
+
+For example you can use these predicates to validate dates straight out of the box:
+
+```ruby
+required(:attr).filled(:date?, lteq?: start_date, gteq?: end_date)
+```
+
+### 2.9 presence
+
+dry-validation has no exact equivalent of ActiveModel's `presence` validation, as in `validates :attr, presence: true`. A first approximation would be `required(:attr).filled`; however there are a few differences.
+
+Internally, ActiveModel's `presence` validation calls the method `present?` on the validated attribute, which is equivalent to `!blank?`. Neither `present?` nor `blank?` are a inbuilt Ruby methods, but a monkey-patch added to every object by ActiveSupport, with the following semantics:
+
+- `nil` and `false` are blank
+- strings composed only of whitespace are blank
+- empty arrays and hashes are blank
+- any object that responds to `empty?` and is empty is blank.
+- everything else is present.
+
+dry-validation's `filled?` predicate is simpler than this, and considers everything to be filled except `nil`, empty Strings, empty Arrays, and empty Hashes.
+
+If you want to validate that a string key contains non-whitespace characters (like ActiveSupport's `String#present?`, you can use a custom predicate such as:
+
+```ruby
+WHITESPACE_PATTERN = /\A[[:space:]#{"\u200B\u200C\u200D\u2060\uFEFF"}]*\z/
+
+def non_blank?(input)
+  !(WHITESPACE_PATTERN =~ input)
+end
+```
+
+**Associations**
+
+If you want to be sure that an association is present, you'll need to create a custom predicate to test whether the associated object itself is present.
+
+If you want to replicate ActiveModel's presence validation of an object associated via a has_one or has_many relationship (checking `.blank?` and `.marked_for_destruction?`), you will need a custom predicate.
+
+**Booleans**
+
+If you want to validate the presence of a boolean field (e.g. true or false) you should use the built in predicate `.bool?`.
+E.g. `required(:attr).filled(:bool?)`
+
+### 2.10 absence
+
+**ActiveModel Validation**
+
+```ruby
+validates :attr, absence: true
+```
+
+**dry-validation**
+
+Dry validation includes two predicates (`empty?` and `none?`) for absence. You should use whichever is most applicable to your situation, remembering that an empty string can be turned into nil using `to_nil` coercion.
+
+```ruby
+required(:attr) { none? }  # only allows nil
```

Comment:
This could use `value` macro so `required(:attr).value(:none?)` - in general we should be using macros for simple conjunctions like this. Blocks should be used only for more complex rules that need ANDs ORs etc.